shared/src/ledger/signing.rs

```diff
@@ -209,24 +195,83 @@ pub async fn sign_tx<U: WalletUtils>(
     wallet: &mut Wallet<U>,
     tx: &mut Tx,
     args: &args::Tx,
-    keypair: &common::PublicKey,
+    public_keys_index_map: &AccountPublicKeysMap,
+    public_keys: &[common::PublicKey],
+    threshold: u8,
 ) -> Result<(), Error> {
-    let keypair = find_key_by_pk(wallet, args, keypair)?;
+    let keypairs = public_keys
+        .iter()
+        .filter_map(|public_key| {
+            match find_key_by_pk(wallet, args, public_key) {
+                Ok(secret_key) => Some(secret_key),
+                Err(_) => None,
+            }
+        })
+        .collect::<Vec<common::SecretKey>>();
+
     // Sign over the transacttion data
-    tx.add_section(Section::Signature(Signature::new(
+    let multisignature_section = MultiSignature::new(
         vec![*tx.data_sechash(), *tx.code_sechash()],
-        &keypair,
-    )));
+        &keypairs,
+        public_keys_index_map,
+    );
+
+    if (multisignature_section.total_signatures() < threshold) && !args.force {
+        return Err(Error::MissingSigningKeys(
+            threshold,
+            multisignature_section.total_signatures(),
+        ));
+    }
+
+    // Sign over the transaction targets
+    tx.add_section(Section::SectionSignature(multisignature_section));
+
     // Remove all the sensitive sections
     tx.protocol_filter();
-    // Then sign over the bound wrapper
+
+    let fee_payer = match &args.fee_payer {
+        Some(keypair) => keypair,
+        None => {
+            if let Some(public_key) = keypairs.get(0) {
+                public_key
+            } else {
+                return Err(Error::InvalidFeePayer(
+                    "Either --signing-keys or --fee-payer must be available."
+                        .to_string(),
+                ));
+            }
+        }
+    };
+
     tx.add_section(Section::Signature(Signature::new(
         tx.sechashes(),
-        &keypair,
+        fee_payer,
     )));
+
     Ok(())
 }
 
+/// Return the necessary data regarding an account to be able to generate a
+/// multisignature section
+pub async fn aux_signing_data<C: crate::ledger::queries::Client + Sync>(
+    client: &C,
+    owner: Option<Address>,
+    public_keys: Vec<common::PublicKey>,
+) -> (AccountPublicKeysMap, u8) {
```

Comment:
I think this function is not very clear - the default threshold `0` when there's no owner doesn't seem right, is it? In e.g. `init_account` that doesn't have an owner beforehand it should respect the tx arg.

I suggest that we change this and use it for all txs except for `reveal_pk`, `init_account` and `init_validator` to only take `owner: &Address` without the `public_keys` and fail if the account info cannot be queried.

For the `reveal_pk`, `init_account` and `init_validator`, we can call `AccountPublicKeysMap::from_iter(public_keys)` directly and take the threshold from the tx arg (with the logic where we set it to 1 when the length of the keys == 1 for new accounts and 1 for `reveal_pk`)

Reply:
thank you!